aptos-move/move-examples/token_v2/sources/common.move

```diff
@@ -0,0 +1,513 @@
+        assert!(amount > 0, error::invalid_argument(EZERO_AMOUNT));
+        assert_burn_cap_and_asset_match(cap, asset);
+        let supply = borrow_supply_mut(asset);
+        assert!(supply.current >= amount, error::invalid_argument(ECURRENT_SUPPLY_UNDERFLOW));
+        supply.current = supply.current - amount;
+    }
+
+    /// ================================================================================================================
+    /// Capability functions
+    /// ================================================================================================================
+    struct MintCap has store {
+        asset_addr: address
+    }
+
+    struct FreezeCap has store {
+        asset_addr: address
+    }
+
+    struct BurnCap has store {
+        asset_addr: address
+    }
+
+    struct asset_owner_caps has drop, store {
+        mint: Option<MintCap>,
+        freeze: Option<FreezeCap>,
+        burn: Option<BurnCap>,
+    }
+
+    public fun new_asset_owner_caps(asset_addr: address, cap_flags: vector<bool>): asset_owner_caps {
+        assert_flags_length(&cap_flags);
+        assert!(*vector::borrow(&cap_flags, 0), error::invalid_argument(ECAP_FLAGS_WITH_NO_MINT));
+        let enable_freeze = *vector::borrow(&cap_flags, 1);
+        let enable_burn = *vector::borrow(&cap_flags, 2);
+        asset_owner_caps {
+            mint: option::some(MintCap { asset_addr }),
+            freeze: if (enable_freeze) { option::some(FreezeCap { asset_addr }) } else { option::none() },
+            burn: if (enable_burn) { option::some(BurnCap { asset_addr }) } else { option::none() }
+        }
+    }
+
+    inline fun borrow_asset_owner_caps<T: key>(asset: &Object<T>): &asset_owner_caps acquires FungibleAssetMetadata {
+        assert_fungible_asset_metadata_exists(asset);
+        &borrow_global<FungibleAssetMetadata>(object_address(asset)).asset_owner_caps
+    }
+
+    inline fun borrow_asset_owner_caps_mut<T: key>(
+        owner: &signer,
+        asset: &Object<T>
+    ): &mut asset_owner_caps acquires FungibleAssetMetadata {
+        assert_owner(owner, asset);
+        assert_fungible_asset_metadata_exists(asset);
+        &mut borrow_global_mut<FungibleAssetMetadata>(object_address(asset)).asset_owner_caps
+    }
+
+    public fun asset_owner_caps_contain_mint<T: key>(asset: &Object<T>): bool acquires FungibleAssetMetadata {
+        option::is_some(&borrow_asset_owner_caps(asset).mint)
+    }
+
+    public fun asset_owner_caps_contain_freeze<T: key>(asset: &Object<T>): bool acquires FungibleAssetMetadata {
+        option::is_some(&borrow_asset_owner_caps(asset).freeze)
+    }
+
+    public fun asset_owner_caps_contain_burn<T: key>(asset: &Object<T>): bool acquires FungibleAssetMetadata {
+        option::is_some(&borrow_asset_owner_caps(asset).burn)
+    }
+
+    public fun get_mint_from_asset_owner_caps<T: key>(
```

Comment:
This allows people to take the MintCap and store in their own module/resource. Is this necessary? Similar question for burn/freeze